scripts/casefold_db.py

```diff
@@ -0,0 +1,285 @@
+    associations: Dict[str, List[Tuple[str, str, str]]] = {}
+
+    # iterate through selected associations, casefold email, rehash it, and add to
+    # associations dict
+    for address, mxid in res.fetchall():
+        casefold_address = address.casefold()
+
+        # rehash email since hashes are case-sensitive
+        lookup_hash = calculate_lookup_hash(sydent, casefold_address)
+
+        if casefold_address in associations:
+            associations[casefold_address].append((address, mxid, lookup_hash))
+        else:
+            associations[casefold_address] = [(address, mxid, lookup_hash)]
+
+    # list of arguments to update db with
+    db_update_args: List[Tuple[str, str, str, str]] = []
+
+    # list of mxids to delete
+    to_delete: List[Tuple[str]] = []
+
+    # list of mxids to send emails to letting them know the mxid has been deleted
+    mxids: List[Tuple[str, str]] = []
+
+    for casefold_address, assoc_tuples in associations.items():
+        db_update_args.append(
+            (
+                casefold_address,
+                assoc_tuples[0][2],
+                assoc_tuples[0][0],
+                assoc_tuples[0][1],
+            )
+        )
+
+        if len(assoc_tuples) > 1:
+            # Iterate over all associations except for the first one, since we've already
+            # processed it.
+            for address, mxid, _ in assoc_tuples[1:]:
+                to_delete.append((address,))
+                mxids.append((mxid, address))
+
+    # iterate through the mxids and send email, let's only send on email per mxid
+    if send_email and not dry_run:
+        for mxid, address in mxids:
+            processed_mxids = []
+
+            if mxid in processed_mxids:
+                continue
+            else:
+                templateFile = sydent.get_branded_template(
+                    "matrix-org",
+                    "migration_template.eml",
+                    ("email", "email.template"),
+                )
+
+                sendEmail(
+                    sydent,
+                    templateFile,
+                    address,
+                    {"mxid": "mxid", "subject_header_value": "MatrixID Update"},
+                )
+                processed_mxids.append(mxid)
+
+    print(
+        f"{len(to_delete)} rows to delete, {len(db_update_args)} rows to update in local_threepid_associations"
+    )
+
+    if not dry_run:
+        if len(to_delete) > 0:
+            cur.executemany(
+                "DELETE FROM local_threepid_associations WHERE address = ?", to_delete
+            )
+
+        if len(db_update_args) > 0:
+            cur.executemany(
+                "UPDATE local_threepid_associations SET address = ?, lookup_hash = ? WHERE address = ? AND mxid = ?",
+                db_update_args,
+            )
+
+        # We've finished updating the database, committing the transaction.
+        db.commit()
+
+
+def update_global_assoc(
```

Comment:
Could you rename this function into `update_global_associations` for consistency please